Volume Accounting Interface methods

pwittrock opened the pull request:
WIP / RFC for methods to add Volume accounting support.

pwittrock committed:
Volume Metrics Interface and base implementation.
- Add volume.MetricsProvider function to Volume interface.
- Add volume.MetricsDu for providing metrics via executing "du".
- Add volulme.MetricsNil for unsupported Volumes.

pwittrock commented:
@vishh

vishh commented:
cc @thockin

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,22 @@ import (
+const unknownSize = -1
+
+type Accounting struct {
+	BytesUsed int
```
What about capacity?

vishh replied:
Should the units be uint64?

pwittrock replied:
Done.

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,22 @@ import (
+	GetAccounting() (*Accounting, error)
+}
+
+const unknownSize = -1
```
Is it possible to avoid this special value?

pwittrock replied:
Yeah I am glad you commented on this.  Is 0 fine for uninitialized, or will we need to ever return 0 as an actual value and differentiate.

thockin replied:
0 seems like a valid size.  Comments please

vishh replied:
Yeah `0` is a valid size. We can use a special error instead.

pwittrock replied:
Done

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,22 @@ import (
+}
+
+type VolumeDefaults struct {
+	AccountingNil
```
Instead of returning `-1`, should we instead return an established error message?

pwittrock replied:
SGTM.  Is this an error condition right now, or should we consider accounting "best effort" and return a nil-error with an uninitialized struct for unsupported volumes?

pwittrock replied:
Done

vishh commented on pkg/volume/accounting_du.go:
```diff
@@ -0,0 +1,50 @@
+	if ad.path == "" {
+		return &Accounting{BytesUsed: unknownSize}, errors.New("No volume path defined for disk usage accounting.")
+	}
+	out, err := exec.Command("du", "-s", ad.path).CombinedOutput()
```
To constrain `du`'s resource usage, we increased its nice value in cadvisor. Relevant code [here](https://github.com/google/cadvisor/blob/master/fs/fs.go#L308).
We should probably start with the same optimization.

pwittrock replied:
Done

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
+
+type Accounting struct {
+	PodBytesUsed    int
+	SharedBytesUsed int
```
What do these fields mean?

pwittrock replied:
PodBytesUsed is the bytes used by the Pod - calculated here by running du -s on the directory
SharedBytesUsed is the bytes used on the actual device.  Perhaps this isn't useful or is something we won't want to support long term and I should remove it.  Thoughts?

thockin replied:
I think the point was to document them very clearly :)

pwittrock replied:
Done

vishh commented:
Just a few comments. Overall structure LGTM. 
ping @saad-ali

thockin commented on pkg/volume/empty_dir/empty_dir.go:
```diff
@@ -103,6 +105,7 @@ func (plugin *emptyDirPlugin) newCleanerInternal(volName string, podUID types.UI
 		mountDetector: mountDetector,
 		plugin:        plugin,
 	}
+	ed.AccountingDu.Init(ed.GetPath())
```
embedding when you have a really generic name like Init() exposed is bound to cause problems.  Maybe InitAccounting()?

pwittrock replied:
Done

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
+	Accountable
+}
+
+type VolumeDefaults struct {
```
Do we need this indirection?  Do we really think there will be other defaults that all have the same overlap to justify a shortcut struct like this?  I'd rather see AccountingNil in each Volume than this.

pwittrock replied:
Nope.  Done.

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
+	AccountingNil
+}
+
+type Accountable interface {
```
commmmmmmments please

pwittrock replied:
Done

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
 	// GetPath returns the directory path the volume is mounted to.
 	GetPath() string
+
+	Accountable
```
comments

pwittrock replied:
Done

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
+
+const unknownSize = -1
+
+type Accounting struct {
```
comments

pwittrock replied:
Done

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
+const unknownSize = -1
+
+type Accounting struct {
+	PodBytesUsed    int
```
int64?  int is not always 64 bits and 4 GiB is not very much these days.

vishh replied:
+1

pwittrock replied:
Done

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,32 @@ import (
+
+func NewAccounting() *Accounting {
+	return &Accounting{
+		PodBytesUsed:    unknownSize,
```
Do you really need to differentiate unknown from 0?

pwittrock replied:
Nope.  Done

pwittrock commented:
I agree will all of the comments, and believe I have addressed them.
Other changes:
- Changed the names of several structs / interfaces / functions to be what I hope are more descriptive.  Open to suggestions on what these should be.
- Added testing for the Metrics implementations
- Changed secret to use MetricsNil until we know that using MetricsDu is the correct thing to do.

thockin commented on pkg/volume/empty_dir/empty_dir.go:
```diff
@@ -133,6 +136,7 @@ type emptyDir struct {
 	mountDetector mountDetector
 	plugin        *emptyDirPlugin
 	rootContext   string
+	volume.MetricsDu
```
minor idea: rather than embedding MetricsDu and then calling Init(), why not embed `MetricsProvider` and initialize it in the struct definition as `MetricsProvider: volume.NewMetricsDu(path)` ?

vishh replied:
+1

pwittrock replied:
Done. Performed some refactoring to get the path value without an initialized emptyDir.  Also made metricsDu unexported so that it cannot be directly created.  Let me know if you prefer unexported metricsDu, or as it was before.

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,28 @@ import (
+}
+
+// VolumeMetricsProvider exposes metrics (e.g. capacity) related to a Volume.
+type VolumeMetricsProvider interface {
```
don't stutter - `volume.MetricsProvider` is better than `volume.VolumeMetricsProvider`

pwittrock replied:
Done

thockin commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,28 @@ import (
+	// Note: For block devices this maybe more than the total size of the files.  See `man du` for details.
+	VolumeBytesUsed uint64
+
+	// FileSystemBytesUsed represents the total bytes used on the filesystem that the Volume path resides.
```
What does this mean for things like emptydir which share a filesystem with the host?  Clarify comments

pwittrock replied:
Renamed fields per Vish's suggestion and used a lot of his language in the comments.  PTAL.

vishh commented on pkg/volume/metrics_du.go:
```diff
@@ -0,0 +1,129 @@
+// See VolumeMetricsProvider.GetCapacityMetrics
+// GetCapacityMetrics calculates the volume usage and device free space by executing "du"
+// and "df" on path.
+func (cm *MetricsDu) GetCapacityMetrics() (*CapacityMetrics, error) {
```
If we were to drop the keyword `Capacity`, the same interface can expose other metrics in the future. WDYT?

pwittrock replied:
The interface doesn't have `Capacity` in it now, but I can drop it from the method / struct.  Wasn't sure if we would want to break metrics of other types out separately or bundle them together.  I'll update it per your suggestion and see how it looks.

pwittrock replied:
Done.

vishh commented on pkg/volume/metrics_du.go:
```diff
@@ -0,0 +1,129 @@
+
+// runDf executes the "df" command and writes the results to metrics.DeviceBytesUsed
+// and metrics.DeviceBytesAvailable
+// TODO(pwittrock): Consider using a statfs syscall directly to get the data
```
I'd prefer switching to statfs right away. For some volumes statfs can be used to get both capacity and usage. So the statfs logic should be abstracted appropriately.

pwittrock replied:
Since statfs is a POSIX call, it breaks the windows build of k8s.  I since cadvisor only builds for linux anyway, I am not sure that there is any benefit in building k8s for windows and may be we should just remove it.  This is causing the Jenkins GCE e2e tests to fail to build.

pwittrock replied:
I created build specific utilities for getting the fsinfo and put them in the util package.  Do you agree that this is the best place?

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +28,28 @@ import (
+type CapacityMetrics struct {
+	// VolumeBytesUsed represents the total bytes used under the directory path that the Volume is mounted.
+	// Note: For block devices this maybe more than the total size of the files.  See `man du` for details.
+	VolumeBytesUsed uint64
```
Assuming that we will follow similar naming conventions for the kubelet metrics APIs, I prefer not exposing filesystem specific concepts.
How about exposing 
1. `Capacity` - The overall capacity of a volume. In the case of shared storage volumes, this will be size of the underlying storage.
2. `Usage` - Usage of a volume. 
3. `Free` - Storage space available for a volume. In the case of shared storage volumes, `Free` will not be equal to `Capacity` - `Usage`.

pwittrock replied:
Done.

pwittrock commented:
@k8s-bot test this

vishh commented:
Is this PR ready for another review @pwittrock?

On Wed, Dec 9, 2015 at 2:46 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e test build/test *passed* for commit 1fde1d2
> <https://github.com/kubernetes/kubernetes/commit/1fde1d2a799f726845708943795b88a84ff60039>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/1fde1d2a799f726845708943795b88a84ff60039/kubernetes-pull-build-test-e2e-gce/19879/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/1fde1d2a799f726845708943795b88a84ff60039/kubernetes-pull-build-test-e2e-gce/19879/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//19879>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18232#issuecomment-163425263>
> .
>

pwittrock commented:
Yes, thanks.  PTAL

thockin commented:
No issues on the volume API

vishh commented on pkg/volume/metrics_nil.go:
```diff
@@ -0,0 +1,31 @@
+// See MetricsProvider.GetMetrics
+// GetMetrics returns an empty Metrics and an error.
+func (*MetricsNil) GetMetrics() (*Metrics, error) {
+	return &Metrics{}, errors.New("metrics are not supported for MetricsNil Volumes")
```
I suspect we might have to expose this error message as a special variable to differentiate between metrics errors vs unsupported scenarios. This code is Ok for this PR though.

pwittrock replied:
Sounds good.  The correct behavior may become more clear once these are actually used.

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +29,32 @@ import (
+
+// MetricsProvider exposes metrics (e.g. used,available space) related to a Volume.
+type MetricsProvider interface {
+	// GetMetrics returns the Metrics for this Volume.  Maybe expensive for some implementations.
```
nit: `for **this** Volume` doesn't read well.

pwittrock replied:
Done

vishh commented on pkg/volume/volume.go:
```diff
@@ -28,6 +29,32 @@ import (
+// Metrics represents the used and available bytes of the Volume.
+type Metrics struct {
+	// Used represents the total bytes used by the Volume.
+	// Note: For block devices this maybe more than the total size of the files.  See `man du` for details.
```
Should we specify `du` here? What if we switch to quota on zfs for example?

pwittrock replied:
Done

vishh commented:
@pwittrock: Just a few nits. Otherwise LGTM. This is no longer a WIP :)

pwittrock commented:
Comments addressed.  PTAL.  I squashed all of the commits into 1.

pwittrock commented:
@k8s-bot test this

rootfs commented on pkg/volume/empty_dir/empty_dir_test.go:
```diff
@@ -274,3 +274,42 @@ func TestPluginBackCompat(t *testing.T) {
+	if err != nil {
+		t.Errorf("Unexpected error when calling GetMetrics %v", err)
+	}
+	if metrics.Used.Value() != 4096 {
```
this is very fs type dependent. In tmpfs, directory size is 0, and thus we'd expect != 4096 is correct. A fs has block size more than 4096, see [`stat(2)` ](http://linux.die.net/man/2/stat)
